### ISSUE TITLE
Utilise CUDA

### ISSUE THREAD

**Issue by Chris-Pedersen** [Chris-Pedersen/torch_qg#19]

**Comment by Chris-Pedersen** [comment]
Implemented in [b098c97](https://github.com/Chris-Pedersen/torch_qg/commit/b098c97c0789658247adcd1b7be55fc98bbdf908). Testing against pyqg:

1. We see flat scaling up until nx=~300, when the GPU fills up.
2. For very small boxes, pyqg is significantly faster, but scales a lot worse. PyQG in the below figure is on 2 cores, running again on 10 cores saw no speedup.
3. Float32 vs float64 makes no difference until the GPU gets full - the GPU fills up sooner with float64, also makes sense.
4. High res sims approx 10x faster on torchqg when using CUDA, than pyqg for same config + numerical scheme, but low res is slower.
5. NB I disabled all diagnostics in both pyqg and torchqg here, such that the comparisons are equal (pyqg has a lot more diagnostics than I'm doing in torchqg so this would be a bit unfair). So we can consider this a conservative bound on the speedup from torchqg

Overall this is cool 
![scaling](https://github.com/Chris-Pedersen/torch_qg/assets/16047009/e9ce3430-bb30-4bc7-8e8e-a0b7f4512dbc)

Still need to put all the diagnostics on torch - currently a mix of numpy and torch which break on gpu.
